src/backend/distributed/executor/distributed_intermediate_results.c

```diff
@@ -0,0 +1,320 @@
+	TupleTableSlot *tupleSlot, DistTableCacheEntry *targetRelation);
+static Tuplestorestate * ExecuteSelectTasksIntoTupleStore(List *taskList, TupleDesc
+														  resultDescriptor);
+
+
+/*
+ * PartitionTasklistResults executes the given task list, and partitions results
+ * of each task based on targetRelation's distribution method and intervals.
+ * Each of the result partitions are stored in the node where task was executed,
+ * and are named as $resultIdPrefix_from_$sourceShardId_to_$targetShardIndex.
+ *
+ * Result is list of DistributedResultFragment, each of which represents a
+ * partition of results. Empty results are omitted. Therefore, if we have N tasks
+ * and target relation has M shards, we will have NxM-(number of empty results)
+ * fragments.
+ */
+List *
+PartitionTasklistResults(char *resultIdPrefix, List *selectTaskList,
+						 DistTableCacheEntry *targetRelation,
+						 bool binaryFormat)
+{
+	/*
+	 * Make sure that this transaction has a distributed transaction ID.
+	 *
+	 * Intermediate results will be stored in a directory that is derived
+	 * from the distributed transaction ID.
+	 */
+	UseCoordinatedTransaction();
+
+	WrapTasksForPartitioning(resultIdPrefix, selectTaskList, targetRelation,
+							 binaryFormat);
+	return ExecutePartitionTaskList(selectTaskList, targetRelation);
+}
+
+
+/*
+ * WrapTasksForPartitioning wraps the query for each of the tasks by a call
+ * to worker_partition_query_result(). Target list of the wrapped query should
+ * match the tuple descriptor in ExecutePartitionTaskList().
+ */
+static void
+WrapTasksForPartitioning(char *resultIdPrefix, List *selectTaskList,
+						 DistTableCacheEntry *targetRelation,
+						 bool binaryFormat)
+{
+	ListCell *taskCell = NULL;
+	ShardInterval **shardIntervalArray = targetRelation->sortedShardIntervalArray;
+	int shardCount = targetRelation->shardIntervalArrayLength;
+
+	ArrayType *minValueArray = NULL;
+	ArrayType *maxValueArray = NULL;
+	Var *partitionColumn = targetRelation->partitionColumn;
+	int partitionColumnIndex = partitionColumn->varoattno - 1;
+	Oid intervalTypeId = partitionColumn->vartype;
+	int32 intervalTypeMod = partitionColumn->vartypmod;
+	Oid intervalTypeOutFunc = InvalidOid;
+	bool intervalTypeVarlena = false;
+	getTypeOutputInfo(intervalTypeId, &intervalTypeOutFunc, &intervalTypeVarlena);
+
+	ShardMinMaxValueArrays(shardIntervalArray, shardCount, intervalTypeOutFunc,
+						   &minValueArray, &maxValueArray);
+	StringInfo minValuesString = ArrayObjectToString(minValueArray, TEXTOID,
+													 intervalTypeMod);
+	StringInfo maxValuesString = ArrayObjectToString(maxValueArray, TEXTOID,
+													 intervalTypeMod);
+
+	foreach(taskCell, selectTaskList)
+	{
+		Task *selectTask = (Task *) lfirst(taskCell);
+		StringInfo wrappedQuery = makeStringInfo();
+		List *shardPlacementList = selectTask->taskPlacementList;
+
+		ShardPlacement *shardPlacement = linitial(shardPlacementList);
```

Comment:
`Assert(list_length(shardPlacementList) == 1)`?

Reply:
Would we need to support replication here? We should at least be able to generate a task per placement and have it execute on whichever placement is available. Not sure about the next step.

Reply:
We may have to return groupId instead of nodeId to be able to use the same query.

Reply:
Added support for replication here in https://github.com/citusdata/citus/pull/3363